### PR TITLE
refactor(i18n): extract Extractor from extract_i18n

### PR DESCRIPTION
**Pull request by mprobst**
I put an extractor into your extract so you can extract while you
extract.

This allows integrators to call Extractor as a library. Also refactors
Extractor a bit so that callers need fewer arguments or arguments that
are at the right semantic level.

The refactoring causes no function change.

### CODE REVIEW

File: modules/@angular/compiler-cli/src/extractor.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+export class Extractor {
+  constructor(
+      private options: tsc.AngularCompilerOptions, private program: ts.Program,
+      public host: ts.CompilerHost, private staticReflector: StaticReflector,
+      private messageBundle: compiler.MessageBundle, private reflectorHost: ReflectorHost,
+      private metadataResolver: compiler.CompileMetadataResolver,
+      private directiveNormalizer: compiler.DirectiveNormalizer) {}
+
+  private readFileMetadata(absSourcePath: string): FileMetadata {
+    const moduleMetadata = this.staticReflector.getModuleMetadata(absSourcePath);
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/src/extractor.ts]: this seems like the place where we should canonicalize the path

Comment (mprobst) [inline modules/@angular/compiler-cli/src/extractor.ts]: Doesn't seem to do it. Let's punt and fix this up in a more general PR.

**Comment by mprobst** [comment]
Test failure is a flake.

**Comment by IgorMinar** [comment]
caretaker note: this pr causes breakages in g3 - @mprobst is working on an internal patch

**Comment by IgorMinar** [comment]
unblocked. CI is red because of the DDOS attack that is still affecting saucelabs, I've seen the job be green in the previous run, so I'm going ahead and merging this.

**Pull request merged**